File: android/truepad-spt/src/main/kotlin/dev/systemslibrarian/truepad/spt/SptRecord.kt
package dev.systemslibrarian.truepad.spt

import dev.systemslibrarian.truepad.core.JsonNumber
import dev.systemslibrarian.truepad.core.JsonObject
import dev.systemslibrarian.truepad.core.JsonString
import dev.systemslibrarian.truepad.core.parseJson

/* ============================================================================
 * The SPT durable record codec — canonical serialization and strict parsing,
 * the Kotlin mirror of the shared validators in the browser engine's state files.
 *
 * Records are tiny JSON objects with a fixed key order. `serialize` produces
 * exactly `{"k":v,...}` as JSON.stringify would; `parseRecord` demands an object
 * of the exact version and the exact key set, so a record copied from one
 * request's directory into another's is rejected, not trusted.
 * ========================================================================= */

const val SPT_RECORD_VERSION = 1

/** A record value is either an Int (only `version`) or an ASCII String. */
internal fun serializeRecord(vararg entries: Pair<String, Any>): ByteArray {
    val sb = StringBuilder("{")
    for ((i, e) in entries.withIndex()) {
        if (i > 0) sb.append(',')
        jsonString(sb, e.first)
        sb.append(':')
        when (val v = e.second) {
            is Int -> sb.append(v.toString())
            is String -> jsonString(sb, v)
            else -> throw IllegalArgumentException("unsupported record value type: ${v::class}")
        }
    }
    sb.append('}')
    return sb.toString().toByteArray(Charsets.UTF_8)
}

/** JSON.stringify-compatible string escaping. The record values are constrained
 *  ASCII (hex/base64url/ISO/enum), but this stays fully correct so a value can
 *  never smuggle an unescaped byte. */
private val HEXC = "0123456789abcdef".toCharArray()

private fun jsonString(sb: StringBuilder, s: String) {
    sb.append('"')
    for (c in s) {
        when (c) {
            '"' -> sb.append("\\\"")
            '\\' -> sb.append("\\\\")
            '\b' -> sb.append("\\b")
            '\u000C' -> sb.append("\\f")
            '\n' -> sb.append("\\n")
            '\r' -> sb.append("\\r")
            '\t' -> sb.append("\\t")
            else -> if (c.code < 0x20) {
                sb.append("\\u00").append(HEXC[(c.code ushr 4) and 0xF]).append(HEXC[c.code and 0xF])
            } else {
                sb.append(c)
            }
        }
    }
    sb.append('"')
}

/** Strict parse: non-empty, valid JSON, an object, the exact record version, and
 *  exactly the given keys (no more, no fewer). Throws on any deviation. */
internal fun parseRecord(bytes: ByteArray, what: String, keys: List<String>): JsonObject {
    if (bytes.isEmpty()) throw IllegalArgumentException("the $what is empty")
    val parsed = try {
        parseJson(String(bytes, Charsets.UTF_8))
    } catch (_: Exception) {
        throw IllegalArgumentException("the $what does not parse as JSON")
    }
    val obj = parsed as? JsonObject ?: throw IllegalArgumentException("the $what is not a JSON object")
    val version = (obj.members["version"] as? JsonNumber)?.raw?.toIntOrNull()
    if (version != SPT_RECORD_VERSION) throw IllegalArgumentException("unsupported $what version")
    val actual = obj.members.keys.sorted()
    val wanted = keys.sorted()
    if (actual != wanted) throw IllegalArgumentException("the $what's fields are wrong")
    return obj
}

internal fun JsonObject.str(key: String): String =
    (members[key] as? JsonString)?.value ?: throw IllegalArgumentException("$key is not a string")

/** base64url field decoded to EXACTLY [length] bytes, canonical spelling required. */
internal fun decodeExact(value: String, length: Int, field: String): ByteArray {
    val bytes = fromBase64Url(value) ?: throw IllegalArgumentException("$field is not canonical unpadded base64url")
    if (bytes.size != length) throw IllegalArgumentException("$field decodes to ${bytes.size} bytes, expected $length")
    if (toBase64Url(bytes) != value) throw IllegalArgumentException("$field has a non-canonical base64url spelling")
    return bytes
}

private val HEX_32_RE = Regex("^[0-9a-f]{32}$")
internal fun isHex32(s: String): Boolean = HEX_32_RE.matches(s)

File: ios/TruePadKit/Sources/TruePadSPT/SptRecord.swift
import TruePadCore

/* ============================================================================
 * The SPT durable record codec — canonical serialization and STRICT parsing.
 *
 * Records are tiny JSON objects with a fixed key order. `serializeRecord`
 * produces exactly `{"k":v,...}` as `JSON.stringify` would; `parseRecord` demands
 * an object of the exact version and the EXACT key set, so a record copied from
 * one request's directory into another's is rejected rather than trusted.
 * ========================================================================= */

public let sptRecordVersion = 1

/// A record value is either an Int (only `version`) or an ASCII String.
enum RecordValue {
    case int(Int)
    case string(String)
}

func serializeRecord(_ entries: [(String, RecordValue)]) -> [UInt8] {
    var s = "{"
    for (i, e) in entries.enumerated() {
        if i > 0 { s.append(",") }
        appendSptJsonString(&s, e.0)
        s.append(":")
        switch e.1 {
        case .int(let v): s.append(String(v))
        case .string(let v): appendSptJsonString(&s, v)
        }
    }
    s.append("}")
    return Array(s.utf8)
}

/// `JSON.stringify`-compatible string escaping.
///
/// The record values are constrained ASCII — hex, base64url, ISO timestamps,
/// enum spellings — but this stays fully correct so a value can never smuggle an
/// unescaped byte into a record another edition will re-parse.
func appendSptJsonString(_ out: inout String, _ s: String) {
    let hexDigits = Array("0123456789abcdef")
    out.append("\"")
    for scalar in s.unicodeScalars {
        switch scalar {
        case "\"": out.append("\\\"")
        case "\\": out.append("\\\\")
        case "\u{08}": out.append("\\b")
        case "\u{0C}": out.append("\\f")
        case "\n": out.append("\\n")
        case "\r": out.append("\\r")
        case "\t": out.append("\\t")
        default:
            if scalar.value < 0x20 {
                out.append("\\u00")
                out.append(hexDigits[Int((scalar.value >> 4) & 0xF)])
                out.append(hexDigits[Int(scalar.value & 0xF)])
            } else {
                out.unicodeScalars.append(scalar)
            }
        }
    }
    out.append("\"")
}

/// Strict parse: non-empty, valid JSON, an object, the exact record version, and
/// EXACTLY the given keys — no more, no fewer. Throws on any deviation.
func parseRecord(_ bytes: [UInt8], what: String, keys: [String]) throws -> [String: JsonValue] {
    if bytes.isEmpty { throw SptRejected(why: "the \(what) is empty") }
    guard let text = String(bytes: bytes, encoding: .utf8),
          let parsed = try? parseStrictJson(text) else {
        throw SptRejected(why: "the \(what) does not parse as JSON")
    }
    guard let members = parsed.memberMap, let actualKeys = parsed.memberKeys else {
        throw SptRejected(why: "the \(what) is not a JSON object")
    }
    guard case .number(let raw)? = members["version"], Int(raw) == sptRecordVersion else {
        throw SptRejected(why: "unsupported \(what) version")
    }
    guard actualKeys.sorted() == keys.sorted() else {
        throw SptRejected(why: "the \(what)'s fields are wrong")
    }
    return members
}

func recordString(_ members: [String: JsonValue], _ key: String) throws -> String {
    guard case .string(let v)? = members[key] else {
        throw SptRejected(why: "\(key) is not a string")
    }
    return v
}

/// A base64url field decoded to EXACTLY `length` bytes, canonical spelling
/// required — a re-encode must reproduce the same text, so padding or an
/// alternative alphabet is refused rather than silently accepted.
func decodeExact(_ value: String, length: Int, field: String) throws -> [UInt8] {
    guard let bytes = SptBytes.fromBase64Url(value) else {
        throw SptRejected(why: "\(field) is not canonical unpadded base64url")
    }
    guard bytes.count == length else {
        throw SptRejected(why: "\(field) decodes to \(bytes.count) bytes, expected \(length)")
    }
    guard SptBytes.toBase64Url(bytes) == value else {
        throw SptRejected(why: "\(field) has a non-canonical base64url spelling")
    }
    return bytes
}

func isSptHex32(_ s: String) -> Bool {
    s.count == 32 && s.allSatisfy { $0.isASCII && ($0.isNumber || ("a"..."f").contains(String($0))) }
}

func isSptHex64(_ s: String) -> Bool {
    s.count == 64 && s.allSatisfy { $0.isASCII && ($0.isNumber || ("a"..."f").contains(String($0))) }
}

func sptHex(_ bytes: [UInt8]) -> String {
    let digits = Array("0123456789abcdef")
    var out = ""
    out.reserveCapacity(bytes.count * 2)
    for b in bytes {
        out.append(digits[Int(b >> 4)])
        out.append(digits[Int(b & 0x0F)])
    }
    return out
}

/* ---- canonical timestamps --------------------------------------------------
 * The SPT durable records store and RE-VALIDATE timestamps in the exact
 * `YYYY-MM-DDTHH:mm:ss.sssZ` spelling. The arithmetic is TruePadCore.IsoTime,
 * shared with the store's bookkeeping so there is one implementation rather than
 * two that can disagree about a leap year — and a disagreement here is a REFUSAL,
 * because these fields are validated by round trip.
 * ------------------------------------------------------------------------- */

public enum SptTime {
    /// Exactly seven days as a duration — NOT "the same clock time seven days
    /// later", which stretches across a DST boundary.
    public static let requestTtlMillis = IsoTime.sevenDaysMillis

    public static func format(epochMillis: Int) -> String { IsoTime.format(epochMillis: epochMillis) }
    public static func isCanonicalIso(_ s: String) -> Bool { IsoTime.isCanonical(s) }
    public static func parseMillis(_ s: String) -> Int? { IsoTime.parseMillis(s) }
}
